[Writing Tools] Quoted text in emails is rewritten (should be excluded from modification)

rr-codes opened the pull request:
#### 4cf517d5ce5eec2f7df269c82e6c6bb24d348265
<pre>
[Writing Tools] Quoted text in emails is rewritten (should be excluded from modification)
<a href="https://bugs.webkit.org/show_bug.cgi?id=276004">https://bugs.webkit.org/show_bug.cgi?id=276004</a>
<a href="https://rdar.apple.com/120563824">rdar://120563824</a>

Reviewed by Abrar Rahman Protyasha.

Add the `WTWritingToolsPreserved` attribute to the ranges corresponding to text enclosed in `blockquote` or `pre` tags,
so that Writing Tools does not rewrite the content.

* Source/WebCore/editing/cocoa/HTMLConverter.h:
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(hasAncestorQualifyingForWritingToolsPreservation):
(WebCore::editingAttributedString):
* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::WritingToolsController::willBeginWritingToolsSession):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, BlockquoteAndPreContentsAreIgnored)):

Canonical link: <a href="https://commits.webkit.org/280542@main">https://commits.webkit.org/280542@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e1b9218847fa82b8ec9a8dcdcade562581294017

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56870 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36198 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9344 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60489 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7313 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58998 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43821 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7503 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46068 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5134 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58900 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34000 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49080 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26926 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30780 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6411 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6317 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52737 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6682 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62171 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/783 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6787 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53319 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/786 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49135 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53354 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/662 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8477 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32027 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33112 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34197 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32858 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -2442,6 +2443,29 @@ AttributedString attributedString(const SimpleRange& range)
 }
 
+#if ENABLE(WRITING_TOOLS)
+static bool hasBlockquoteOrPreAncestor(const Node& node)
```
Nit: doesn't seem like a bad method to have unconditional of `ENABLE(WRITING_TOOLS)`, either 😛

pxlcoder commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -2442,6 +2443,29 @@ AttributedString attributedString(const SimpleRange& range)
+        if (hasBlockquoteOrPreTag)
+            return true;
+
+        element = element->parentElement();
```
I'm still a little concerned about this approach. We're going to be doing a lot of redundant DOM traversal.

rr-codes replied:
I added a caching mechanism

rr-codes commented on Source/WebCore/editing/cocoa/TextIteratorHTMLConverter.mm:
```diff
@@ -0,0 +1,346 @@
+        auto& style = renderer->style();
+
+#if ENABLE(WRITING_TOOLS)
+        // FIXME: Use the `WTWritingToolsPreservedAttributeName` constant instead of a string.
```
need to figure out how to do this

pxlcoder replied:
Would need to soft link Writing Tools in WebCore, not sure it's worth it.

rr-codes commented on Source/WebCore/editing/cocoa/TextIteratorHTMLConverter.mm:
```diff
@@ -0,0 +1,346 @@
+        if (isMailBlockquote(*ancestor))
+            return true;
+
+        // FIXME: This should use the render style of the element instead.
```
need to figure out how to do this

pxlcoder replied:
`renderer()->style()`?

rr-codes replied:
I specifically meant, how can I tell if it is a pre from the render style, not how to access the style.

pxlcoder replied:
whiteSpace() == WhiteSpace::Pre

pxlcoder commented on Source/WebCore/editing/cocoa/TextIteratorHTMLConverter.mm:
```diff
@@ -0,0 +1,357 @@
+    auto fontManager = [NSFontManager sharedFontManager];
+#endif
+
+    HashMap<Element*, bool> m_elementQualifiesForWritingToolsPreservationCache;
```
`WeakHashMap` or `RefPtr<Element>`.

rr-codes replied:
hm, I get this error when using WeakHashMap


```
OpenSource/WebKitBuild/Products/Debug/usr/local/include/wtf/WeakHashMap.h:372:16: error: no viable conversion from returned value of type 'WebCore::WeakPtrImplWithEventTargetData' to function return type 'RefType' (aka 'Ref<WTF::DefaultWeakPtrImpl>')
  372 |         return *key.weakPtrFactory().template createWeakPtr<T>(const_cast<T&>(key)).m_impl;
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
OpenSource/WebKitBuild/Products/Debug/usr/local/include/wtf/WeakHashMap.h:201:36: note: in instantiation of function template specialization 'WTF::WeakHashMap<WebCore::Element, bool>::makeKeyImpl<WebCore::Element>' requested here
  201 |         auto result = m_map.ensure(makeKeyImpl(key), functor);
      |                                    ^
OpenSource/Source/WebCore/editing/cocoa/HTMLConverter.mm:2447:18: note: in instantiation of function template specialization 'WTF::WeakHashMap<WebCore::Element, bool>::ensure<(lambda at OpenSource/Source/WebCore/editing/cocoa/HTMLConverter.mm:2447:36)>' requested here
 2447 |     return cache.ensure(*ancestor, [&] {
 ```

whsieh replied:
I assume you tried using `WeakHashMap<Element, bool, WeakPtrImplWithEventTargetData>>`?

rr-codes replied:
oops! I did end up getting it working, but then it looks like I accidentally lost that change when doing all the un-refactoring

rr-codes replied:
oh wait it is already in the diff

pxlcoder commented on Source/WebCore/editing/cocoa/TextIteratorHTMLConverter.mm:
```diff
@@ -0,0 +1,357 @@
+    auto fontManager = [NSFontManager sharedFontManager];
+#endif
+
+    HashMap<Element*, bool> m_elementQualifiesForWritingToolsPreservationCache;
```
Also `m_` prefix should not be used for a local variable.

pxlcoder commented on Source/WebCore/editing/cocoa/TextIteratorHTMLConverter.mm:
```diff
@@ -0,0 +1,357 @@
+#if ENABLE(WRITING_TOOLS)
+        // FIXME: Use the `WTWritingToolsPreservedAttributeName` constant instead of a string.
+        if (hasAncestorQualifyingForWritingToolsPreservation(node->parentElement(), m_elementQualifiesForWritingToolsPreservationCache))
+            [attrs setObject:@(1) forKey:@"WTWritingToolsPreserved"];
```
Can "WTWritingToolsPreserved" be defined in a variable at the top of the file?

pxlcoder commented on Source/WebCore/editing/cocoa/TextIteratorHTMLConverter.mm:
```diff
@@ -0,0 +1,357 @@
+
+#if ENABLE(WRITING_TOOLS)
+        // FIXME: Use the `WTWritingToolsPreservedAttributeName` constant instead of a string.
+        if (hasAncestorQualifyingForWritingToolsPreservation(node->parentElement(), m_elementQualifiesForWritingToolsPreservationCache))
```
We should add an option used when creating `EditingHTMLConverter`, since this logic shouldn't be run when generating attributed strings for other purposes.

pxlcoder reviewed (commented):
Not a big fan of including the refactor into two files as part of this fix.

rr-codes committed:
[Writing Tools] Quoted text in emails is rewritten (should be excluded from modification)
https://bugs.webkit.org/show_bug.cgi?id=276004
rdar://120563824

Reviewed by Abrar Rahman Protyasha.

Add the `WTWritingToolsPreserved` attribute to the ranges corresponding to text enclosed in `blockquote` or `pre` tags,
so that Writing Tools does not rewrite the content.

* Source/WebCore/editing/cocoa/HTMLConverter.h:
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(hasAncestorQualifyingForWritingToolsPreservation):
(WebCore::editingAttributedString):
* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::WritingToolsController::willBeginWritingToolsSession):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, BlockquoteAndPreContentsAreIgnored)):

Canonical link: https://commits.webkit.org/280542@main

alanbaradlay commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -2434,6 +2438,24 @@ static String preferredFilenameForElement(const HTMLImageElement& element)
 }
 
+#if ENABLE(WRITING_TOOLS)
+static bool hasAncestorQualifyingForWritingToolsPreservation(Element* ancestor, WeakHashMap<Element, bool, WeakPtrImplWithEventTargetData>& cache)
```
Any particular reason why this is not `const Element* ancestor`?
